auth/CognitoSecureStorage.ts

```diff
@@ -36,109 +37,121 @@ const STORE_CHUNK_SIZE = 2048
 /**
  * Encrypted secure storage for AWS Cognito Auth.
  */
-export class CognitoSecureStorage {
-  private syncPromise?: Promise<void>
+export class CognitoSecureStorage implements KeyValueStorageInterface {
   private store: SecureStore
   private keyChunkMappings = new Map<string, number>()
-  private cache = new Map<string, string>()
+  private keyChunkMappingsPromise?: Promise<Map<string, number>>
 
   constructor (store: SecureStore) {
     this.store = store
+    this.keyChunkMappingsPromise = undefined
   }
 
-  setItem (key: string, value: string) {
-    const chunks = stringToChunks(value)
-    this.keyChunkMappings.set(key, chunks.length)
-    this.saveKeyChunkMappings()
-    this.cache.set(key, value)
-    for (let i = 0; i < chunks.length; i++) {
-      this.store.setItemAsync(secureStoreKey(key, i), chunks[i])
+  async removeItem (key: string) {
+    const keyChunkMappings = await this.loadKeyChunkMappings()
+    if (keyChunkMappings) {
+      const amountOfChunks = keyChunkMappings.get(key) ?? 0
+      await Promise.allSettled(
+        ArrayUtils.repeatElements(amountOfChunks, async (chunkIndex) => {
+          await this.store.deleteItemAsync(this.secureStoreKey(key, chunkIndex))
+        })
+      )
+      keyChunkMappings.delete(key)
     }
+    this.keyChunkMappings = keyChunkMappings
+    await this.saveKeyChunkMappings()
   }
 
-  getItem (key: string) {
-    return this.cache.get(key)
+  async setItem (key: string, value: string) {
+    const keyChunkMappings = await this.loadKeyChunkMappings()
+    if (keyChunkMappings) {
+      const chunks = this.stringToChunks(value)
+      for (let i = 0; i < chunks.length; i++) {
+        const chunkKey = this.secureStoreKey(key, i)
+        await this.store.setItemAsync(chunkKey, chunks[i])
+      }
+      keyChunkMappings.set(key, chunks.length)
+      this.keyChunkMappings = keyChunkMappings
+      await this.saveKeyChunkMappings()
+    }
   }
 
-  removeItem (key: string) {
-    this.cache.delete(key)
-
-    const amountOfChunks = this.keyChunkMappings.get(key) ?? 0
-    for (let i = 0; i < amountOfChunks; i++) {
-      this.store.deleteItemAsync(secureStoreKey(key, i))
+  async getItem (key: string) {
+    const keyChunkMappings = await this.loadKeyChunkMappings()
+    if (keyChunkMappings.size > 0) {
+      const amountOfChunks = keyChunkMappings.get(key) ?? 0
+      const chunks = [] as string[]
+      for (let i = 0; i < amountOfChunks; i++) {
+        const chunk = await this.store.getItemAsync(this.secureStoreKey(key, i))
+        if (chunk) {
+          chunks.push(chunk)
+        }
+      }
```

Comment:
This can also be done with `Promise.allSettled` to avoid doing the async work sequentially.